src/main/java/at/hannibal2/skyhanni/utils/StringUtils.kt

```diff
@@ -34,30 +34,61 @@ object StringUtils {
      *
      * @param keepFormatting Boolean indicating whether to retain non-color formatting codes (default: false).
      * @return A string with color codes removed (and optionally formatting codes if specified).
      */
-    fun String.removeColor(keepFormatting: Boolean = false): String {
-        val builder = StringBuilder(this.length)
-
-        var counter = 0
-        while (counter < this.length) {
-            if (this[counter] == '§') {
-                if (!keepFormatting || this[counter + 1] !in formattingChars) {
-                    counter += 2
-                    continue
-                }
+    fun CharSequence.removeColor(keepFormatting: Boolean = false): String {
+        // Glossary:
+        // Formatting indicator: The '§' character indicating the beginning of a formatting sequence
+        // Formatting code: The character following a formatting indicator which specifies what color or text style this sequence corresponds to
+        // Formatting sequence: The combination of a formatting indicator and code that changes the color or format of a string
+
+        // Find the first formatting indicator
+        var nextFormattingSequence = indexOf('§')
+
+        // If this string does not contain any formatting indicators, just return this string directly
+        if (nextFormattingSequence < 0) return this.toString()
+
+        // Let's create a new string, and pre-allocate enough space to store this entire string
+        val cleanedString = StringBuilder(this.length)
+
+        // Read index stores the position in `this` which we have written up until now
+        // a/k/a where we need to start reading from
+        var readIndex = 0
+
+        // As long as there still is a formatting indicator left in our string
+        while (nextFormattingSequence >= 0) {
+
+            // Write everything from the read index up to the next formatting indicator into our clean string
+            cleanedString.append(this, readIndex, nextFormattingSequence)
+
+            // If the next formatting sequence's code indicates a non-color format and we should keep those
+            if (keepFormatting && nextFormattingSequence + 1 < length && this[nextFormattingSequence + 1] in formattingChars) {
+                // Set the readIndex to the formatting indicator, so that the next loop will start writing from that paragraph symbol
+                readIndex = nextFormattingSequence
+                // Find the next § symbol after the formatting sequence
+                nextFormattingSequence = indexOf('§', startIndex = readIndex + 1)
+            } else {
+                // If this formatting sequence should be skipped (either a color code, or !keepFormatting or an incomplete formatting sequence without a code)
+                // Set the readIndex to after this formatting sequence, so that the next loop will skip over it before writing the string
+                readIndex = nextFormattingSequence + 2
```

Comment:
When using the suggestion (and deleting the out of bound check below) there is the possibility to remove a branch. Does not require to use min could simply be a the bound if from below. I'm unsure what is more performant since technically min could be implemented branchless (but unlikely). Either way move the boundscheck inside here, since the other branch can't get out of bounds
```suggestion
                readIndex = min(nextFormattingSequence + 2, this.length)
```

Reply:
`(nextFormattingSequence+2).coerceAtLeast(this.length)`

Reply:
btw. it is coerceAtMost 
But this is literally the wrapper for the boundscheck below. Min *could* be implemented better performant then a branch. And since performance is important here (else we could simply use the old removeColor) we should use the more optimal method for this, any other case I would use coerceAtMost here for readability.

Reply:
> And since performance is important here (else we could simply use the old removeColor) we should use the more optimal method for this, any other case I would use coerceAtMost here for readability.

I have to strongly disagree. While performance is the matter of this PR, there is a difference between significant and measurable improvements and potentially maybe saving time on the order of a few clock cycles at best.

Reply:
I also want to point out here that the compiler is king for small stuff like this. Optimizing and inlining functions like this is often something the compiler is very good at. Very likely that whatever intrinsic Math.min has, there is a similar optimization for a simple tertiary in JIT. In fact i could make an argument (just as bogus, *especially* without benchmarking) that Math.min has worse performance, since calling the function requires overhead for the stack of the Math.min call to be allocated (which is a quite a few instructions). Sure you could inline the function, but even when inlining functions certain attributes about stacktraces must be preserved, meaning that there is still some runtime stacktrace overhead. In practice none of this matters without actual benchmarking.

Reply:
Someone could go this way, but any reasonable person would say no. I suggested the move of the bounds check and also a way to write it (```min```). I do not care if it is a ```min```, a ```coerceAtMost```, a ```if```, a ```trinary``` or some bit operation to get the same result. I would maybe care if I tested and found a definite best but since I don't have any good testing environment and I can't be bothered to setup one for this thing, I do not care. The ```min``` is just what I know as the best way to do it normally (outside of java).

Reply:
Also approve your commit (5582cd0254ac61b92e955f41fee640abd8d2c553). 
And thx for the testing.